lib/Transforms/StagedPatternRewriteDriver.cpp

```diff
@@ -0,0 +1,289 @@
+public:
+  explicit SinglePatternRewriteDriver(ArrayRef<Operation *> roots,
+                                      const FrozenRewritePatternSet &patterns)
+      : PatternRewriter(roots.front()->getContext()), matcher(patterns) {
+    // Apply a simple cost model based solely on pattern benefit.
+    matcher.applyDefaultCostModel();
+    worklist.assign(roots.begin(), roots.end());
+    // TODO: reverse if we need the reverse order.
+    for (size_t i = 0, e = worklist.size(); i != e; ++i)
+      worklistMap[worklist[i]] = i;
+  }
+
+  /// During one local round, we traverse the list of ops that currently need
+  /// processing and we apply at most one pattern that has not yet been applied.
+  LogicalResult doOneLocalRound() {
+    LLVM_DEBUG({
+      logger.getOStream() << "\n";
+      logger.startLine() << logLineComment;
+      logger.startLine() << "Start staged pattern round : " << roundNum
+                         << "\n\n";
+    });
+    ++roundNum;
+
+    if (worklist.empty())
+      return failure();
+
+    // Reset the ops to process to prepare for next round.
+    LogicalResult res = failure();
+
+    // We want to iterate over all the ops to process this round, apply one
+    // pattern then exit.
+    do {
+      Operation *op = popFromWorklist(worklist, worklistMap);
+
+      LLVM_DEBUG({
+        logger.getOStream() << "\n";
+        logger.startLine() << logLineComment;
+        logger.startLine() << "Staged-processing operation : '" << op->getName()
+                           << "'(" << op << ") {\n";
+        logger.indent();
+
+        // If the operation has no regions, just print it here.
+        if (op->getNumRegions() == 0) {
+          op->print(
+              logger.startLine(),
+              OpPrintingFlags().printGenericOpForm().elideLargeElementsAttrs());
+          logger.getOStream() << "\n\n";
+        }
+      });
+
+      // Only apply one pattern of the list.
+      bool appliedThisRound = false;
+      auto canApply = [&](const Pattern &pattern) {
+        bool res = !appliedPatterns.contains(&pattern);
+        LLVM_DEBUG({
+          logger.startLine()
+              << "** canApply  : '" << &pattern << " -> " << res << ")\n";
+        });
+        return res;
+      };
+
+      auto onFailure = [&](const Pattern &pattern) {
+        LLVM_DEBUG({
+          logger.startLine() << "** failed to apply  : '" << &pattern << ")\n";
+        });
+      };
+
+      auto onSuccess = [&](const Pattern &pattern) {
+        appliedThisRound = true;
+        appliedPatterns.insert(&pattern);
+        LLVM_DEBUG(
+            { logger.startLine() << "** Applied  : '" << &pattern << ")\n"; });
+        return success();
+      };
+
+      if (succeeded(matcher.matchAndRewrite(op, *this, canApply, onFailure,
+                                            onSuccess)))
+        res = success();
+
+      if (appliedThisRound)
+        break;
+    } while (!worklist.empty());
```

Comment:
Just to make sure the application logic that I understand from code is consistent with intent:
- if no patterns apply, an op is just removed from the worklist;
- if no patterns apply to any op, the entire application is considered a failure;
- one pattern can only ever apply to one op.

The first two seems reasonable. The last part feels too constraining. I understand we don't want repeated application of a pattern to the ops produced by said pattern, but we likely do want application of a pattern to independent ops that match (otherwise, we would need to insert as many pattern clones as there are ops that match it, doing the matching upfront).

Reply:
I am biased by the test I wrote:

```
// CASE3-LABEL: func @test3() {
func @test3() {
  // CASE3: "start"() {__test_attr__ = 1 : i32} : () -> ()
  // CASE3: "step_1"() : () -> ()
  "start"() : () -> ()
  "start"() : () -> ()
}
```

Basically I wanted the pattern to only apply to the first op but maybe that is not a constraint that should be put on this level of the IR / application hierarchy.

Reply:
what do you think of evolving this after we connect your stuff ?

Reply:
Let's just have a TODO that clearly explains what happens so we know easily where to go when revisiting this. My thinking is already at a point where we target operations more specifically in transformations, "the first op that matches" being only one of the possibilities.

Reply:
Added more doc to the method, thanks!